### DISCUSSION TITLE
Synthesizing a raw s16le waveform

### DISCUSSION THREAD

**Discussion by ghost** [FluidSynth/fluidsynth#952] [Q&A] [answered]
The example presented at https://www.fluidsynth.org/api/fluidsynth_simple_8c-example.html would require other libraries for writing directly on sound drivers. Unfortunately I'm one of those people who posses a potato system and unmaintained packages and libraries for doing such feat. As such, I would like then to store the synthesize samples into a `int16_t *` (preferably) in preparation to my next endeavor.

While I'm aware that fluidsynth can write to stdout, by running `fluidsynth -T raw -n -F /proc/self/fd/1 my.mid`, I question the following:

- How could I set the preset to be use to play?
- How could I set the duration of the note to be played (maybe in number of samples?)
- How could I set the note to be played? (example c3,  d3, e3)
- How can I generate the note with their parameters described above?

For starters, here is what I have done from scrolling through the documentation:
```
/* FluidSynth Simple - An example of using fluidsynth
 *
 * This code is in the public domain.
 *
 * To compile:
 *   gcc -g -O -o fluidsynth_simple fluidsynth_simple.c -lfluidsynth
 *
 * To run
 *   fluidsynth_simple soundfont
 *
 * [Peter Hanappe]
 */
#include <fluidsynth.h>
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <stdint.h>

int main(int argc, char **argv) {
  fluid_settings_t *settings;
  fluid_synth_t *synth = NULL;
  int err = 0;
  if (argc != 2) {
    fprintf(stderr, "Usage: fluidsynth_simple [soundfont]\n");
    return 1;
  }
  /* Create the settings object. This example uses the default
   * values for the settings. */
  settings = new_fluid_settings();
  if (settings == NULL) {
    fprintf(stderr, "Failed to create the settings\n");
    err = 2;
    goto cleanup;
  }
  /* Create the synthesizer */
  synth = new_fluid_synth(settings);
  if (synth == NULL) {
    fprintf(stderr, "Failed to create the synthesizer\n");
    err = 3;
    goto cleanup;
  }
  /* Load the soundfont */
  if (fluid_synth_sfload(synth, argv[1], 1) == -1) {
    fprintf(stderr, "Failed to load the SoundFont\n");
    err = 4;
    goto cleanup;
  }
  /* Play a note */
  fluid_synth_noteon(synth, 0, 60, 100);

  float *left = calloc(sizeof(float), 44100);
  float *right = calloc(sizeof(float), 44100);
  assert(left != NULL && right != NULL);
  float *fx[2], *dry[2];
  fx[0] = left;
  fx[1] = right;
  dry[0] = left;
  dry[1] = right;

  fluid_synth_process(synth, 44100, 2, fx, 2, dry);

  for (int i = 0; i < 44100; i++)
    fwrite(&(int16_t){(int16_t)dry[0][i]}, 1, sizeof(int16_t), stdout);

  fflush(stdout);
cleanup:
  if (synth) {
    delete_fluid_synth(synth);
  }
  if (settings) {
    delete_fluid_settings(settings);
  }
  free(left); free(right);
  return err;
}
```
Running the following commands:
```
$ ./a.out $PREFIX/share/soundfonts/default.sf2
$ echo $?
0
$ ./a.out $PREFIX/share/soundfonts/default.sf2 | hexdump -C
00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00015880
$
```

I suppose `fluid_synth_process(synth, 44100, 2, fx, 2, dry);` did nothing here as the output are all zeros.

**Comment by derselbst** [comment, marked as answer]
> How could I set the preset to be use to play?

e.g. `fluid_synth_program_select()`; but usually, it's up to the .mid file to make this selection.

> How could I set the duration of the note to be played (maybe in number of samples?)

`fluid_synth_noteon()` and `fluid_synth_noteoff()` take effect immediately after rendering audio. Thus it's up to you to render "as-many-samples-as-you-want" before calling `fluid_synth_noteoff()`.

> I suppose fluid_synth_process(synth, 44100, 2, fx, 2, dry); did nothing here as the output are all zeros.

`fluid_synth_process()` outputs float samples. Float samples are in the range [-1;1]. Int16 samples are in the range [-32768;32767]. Your code just casts all the float samples to int16 without proper conversion. Also, consider to use `fluid_synth_write_s16()`.

**Reply by ghost** [reply]
Finally I hear some sounds!! I wonder why this normalization of the float wasn't written on the doc though.

**Reply by derselbst** [reply]
It's a de-facto standard in the audio world that float samples are normalized. Just as it is for int16 samples to use the full range. So  I do wonder where exactly in the docs you would have expected it to be mentioned?

**Reply by ghost** [reply]
I did saw this normalization on other software like the sndfilter but didn't expect it to be widely used:

https://github.com/velipso/sndfilter/blob/bc32dba43af781643eaf436eadcc63120fae882f/src/wav.c#L126-L133

I thought it was just the author's convention for representing this samples.
